File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        res = []

        def backtrack(start,  currSum, curr):

            #Base Case
            if currSum == target:
                res.append(curr.copy())
                return 
            
            #currSUm > target. RETURN. THIS DOES NOT WORK. 
            if currSum > target:
                return

            #Recursive backtracking..
            for i in range(start, len(candidates)):
                curr.append(candidates[i])

                #Call our function to backtrack
                backtrack(i, currSum + candidates[i], curr)

                #Backtrack...
                curr.pop()
        

        backtrack(0, 0, [])

        return res

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> curr;
        backtrack(0, 0, curr, candidates, target);

        return res;
    }

    void backtrack(int i, int currSum, vector<int>& curr, vector<int>& candidates, int target)
    {

        if (currSum > target)
        {
            return;
        }

        if (currSum == target)
        {
            res.push_back(curr);
            return;
        }

        for (int j = i; j < candidates.size(); j++)
        {

            curr.push_back(candidates[j]);
            backtrack(j, currSum + candidates[j], curr, candidates, target);
            curr.pop_back();
        }

        return;
    }

private:

    vector<vector<int>> res;
};